blocks/autocompleters/index.js

```diff
@@ -51,55 +63,52 @@ import BlockIcon from '../block-icon';
  * @typedef {Object} Completer
  * @property {?String} className A class to apply to the popup menu.
  * @property {String} triggerPrefix the prefix that will display the menu.
- * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {FnGetOptions} getOptions get the completer options in a resolved promise.
+ * @property {?FnGetOptionKeywords} getOptionKeywords get the keywords for a given option.
+ * @property {FnRenderOptionLabel} renderOptionLabel render the label for a given option.
  * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
  * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
- * @property {FnOnSelect} onSelect
+ * @property {FnGetOptionResult} getOptionResult
  */
 
 /**
  * Returns an "completer" definition for selecting from available blocks to replace the current one.
  * The definition can be understood by the Autocomplete component.
  *
- * @param {Function} onReplace  Callback to replace the current block.
- *
  * @return {Completer} Completer object used by the Autocomplete component.
  */
-export function blockAutocompleter( { onReplace } ) {
-	// Prioritize common category in block type options
-	const options = sortBy(
-		getBlockTypes(),
-		( { category } ) => 'common' !== category
-	).map( ( blockType ) => {
-		const { name, title, icon, keywords = [] } = blockType;
-		return {
-			value: name,
-			label: [
-				<BlockIcon key="icon" icon={ icon } />,
-				title,
-			],
-			keywords: [ ...keywords, title ],
-		};
-	} );
-
-	const getOptions = () => Promise.resolve( options );
+export function blockAutocompleter() {
+	const getOptions = once( () => Promise.resolve(
+		// Prioritize common category in block type options
+		sortBy(
+			getBlockTypes(),
+			( { category } ) => 'common' !== category
+		)
+	) );
+	const getOptionKeywords = ( { title, keywords = [] } ) => [ ...keywords, title ];
+	const renderOptionLabel = ( { icon, title } ) => [
+		<BlockIcon key="icon" icon={ icon } />,
+		title,
+	];
 
 	const allowContext = ( before, after ) => {
 		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
 	};
 
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
+	const getOptionResult = blockData => createBlock( blockData.name );
 
 	return {
+		type: 'block',
 		className: 'blocks-autocompleters__block',
 		triggerPrefix: '/',
 		getOptions,
+		getOptionKeywords,
+		renderOptionLabel,
```

Comment:
The reason is that they are separate concerns. Otherwise, if you want to override label rendering, you would also have to take responsibility for option retrieval and keywords. This way, you can override retrieval, rendering, and keywords separately.

Reply:
I think I have a small prefernece for the old API here, it looked simpler and easier to understand. 
Is it really a common use-case to replace just one part of the autocompleter? My thinking is that the autocompleters are simple to write and in general we would provide new ones instead of replacing a property? (even if there's some duplication)

Reply:
> I think I have a small prefernece for the old API here, it looked simpler and easier to understand.

I'm sympathetic to this but am thinking it is better to separate concerns. Currently, if a developer wants to simply tweak an option label, they also have to take responsibility for retrieving options and providing option keywords. It requires that they know and understand what the default behavior so they can provide an default-like experience that only overrides the label. If the default changes and they do not notice, the custom completer will unintentionally differ from the default.

Since `getOptions` requires the implementor to think about providing a rendered `label` and `keywords` in addition to option retrieval, I'm not sure it is more complex in practice than having dedicated functions for label and keywords. They are still things that must be considered, but with the updated interface, they can be overridden and tested separately.

I'm not sure I've done a great job at selling this, so any more thoughts here?

Reply:
I understand the concerns but I guess the answer to this question would give us the best option here.

> Is it really a common use-case to replace just one part of the autocompleter?

I personally don't think it's common use-case but I may be wrong. And it's not a blocker for me anyway in this PR :)